### ISSUE TITLE
what is `stats`?

### ISSUE THREAD

**Issue by twatteyne** [openschc/openschc#119]
@MarinoMtz I have a hard time understanding what `stats/` is. Given that there is an `__init__.py`, I assume it's package which can be imported. I'm suprised to see `log.txt` which appears to be an artifact. I do see that it is imported a bit everywhere. I would recommend to, at least:
- rename, as "stats" is super super generic
- add a `README.md` to explain what it is and how it is used

**Comment by MarinoMtz** [comment]
As far as I know, these files are created after launching a simulation to store the logs, I do not use it personally, I think we should delete it and keep it internal for the user, I should be be added to .ignore imho, what do you think @adjih ?

**Comment by twatteyne** [comment]
@MarinoMtz I understand you refer to file `log.txt` and `log2.txt`? https://github.com/openschc/openschc/tree/develop/src/stats also contain quite a bit of Python code. @adjih , any input? Based on your input, we could remove, or merge into a single file.

**Comment by MarinoMtz** [comment]
Yes, that's it, At first I thought since there are some Time On air, it was used to simulate collisions, but I don't see how it's done inside.

**Comment by adjih** [comment]
The code was by Sergio Aguilar. I think it might be the code that was used in the article
[Performance Analysis and Optimal Tuning of IETF LPWAN SCHC ACK-on-Error Mode](https://doi.org/10.1109/JSEN.2020.3007855).
 It is associated with the file https://github.com/openschc/openschc/blob/master/src/test_statsct.py and https://github.com/openschc/openschc/blob/master/src/test_statsct_ack_always.py .
Because this was probably the most extensive use of OpenSCHC for fragmentation with various options, I think some care was taken to keep it running at that various points in time.

**Comment by MarinoMtz** [comment]
Coud you explain us how this was done @saguilarDevel ?

**Comment by adjih** [comment]
It is used a little bit everywhere, to record what happens to the packets and to collect statistics (it would be a little bit equivalent to ns3 simulator). At the end, it is used to print statistics of the simulation. 
It might even be enabled by default, as it is used for instance in `frag_send.py` .

It is used as a module. It would be not so easy to remove it for now (and it is useful); and it would be much easier to bring the equivalent of`test_statsct.py` or `test_statsct_ack_always.py` back, first to do changes.

I suggest: 
* it can be left as is, 
* or  renamed (refactored: updating all the code, including examples/)
  * `src/stat` -> `src/sim_stats` for instance
  * or move `src/stats/statsct.py` into `src/net_sim_stats.py' for instance and move all the content (that is the two functions) of `src/cdf_calc.py`, `src/toa_calculator.py` inside that file as well.
* the logs can be removed (we can no longer generate the same logs, and I don't see the post-processing script anymore)

**Comment by twatteyne** [comment]
@adjih very clear guidance, thanks! I believe it makes sense to move the code into a single file (your option 2).

What you mean  by "bring back" in?

> it would be much easier to bring the equivalent oftest_statsct.py or test_statsct_ack_always.py back, first to do changes.

**Comment by twatteyne** [comment]
@adjih very clear guidance, thanks! I believe it makes sense to move the code into a single file (your option 2).

What you mean  by "bring back" in?

> it would be much easier to bring the equivalent oftest_statsct.py or test_statsct_ack_always.py back, first to do changes.

**Comment by adjih** [comment]
"bring back" = make it run again. These are the full use cases for `stats/`.

`src/test_frag_new.py`,  `src/test_statsct.py`, `src/test_statsct_ack_always.py`, `src/test_newschc.py` were also more or less derived from the same ancestor, implementing the base scenario of two nodes doing fragmentation in a simulator. But  `src/test_statsct.py`, `src/test_statsct_ack_always.py` have not been maintained recently in sync even in the branch `master`.
 
On the other hand, they are just adding a bunch of statistics print to a simulations.

> What you mean by "bring back" in?
> 
> > it would be much easier to bring the equivalent oftest_statsct.py or test_statsct_ack_always.py back, first to do changes.

**Comment by saguilarDevel** [comment]
Hello and sorry the delay response.
Yes, this module was build at the beginning, when OpenSCHC was using micropython, and the results obtained were used for this article [https://link.springer.com/chapter/10.1007/978-3-030-31831-4_15](url). The idea is that you can collect information from every transmission and obtained averages from all the transmissions performed. The ACK-Always is a kind of patch, as it uses ACK-on-Error multiple times to simulate the ACK-Always behavior. 
The collision simulator is based in the idea of a time line of transmission, were you check if you can transmit a fragment or if the channel is busy. This was build to see the effect of collision and evaluate SCHC with this kind of error pattern, however I am not sure how functional it is.
I thought this was in another brach, as there was an IETF Hackathon were all code was migrate to python and refactor completely, so I continue using this in a separate brach and then I clone it and continue working here [https://github.com/saguilarDevel/open_schc](url). I am not sure how different are the branches, or in what branch you are working. 
If you let me know in what branches you are working on and I can check if there are any newer code I added and see if we can merge it, or see how different it is.
I think this kind of module is useful for testing and research, as it gives statistics of many parameters of the SCHC Packet transfer.

**Comment by twatteyne** [comment]
Thanks all for the context!!

@saguilarDevel , we are working on "stabilizing" the code: we want OpenSCHC to be super easy to use, intuitive to understand, targeting any networking engineer/researcher interested in trying out SCHC compression. We are working on branch `develop`. As part of that, we are going through the code, removing what isn't used, and documenting/reworking most of the code. We are looking to either remove or fully integrate different patches/addons people have been contributing if these were only used for a paper or a hackathon, but are not used anymore now and are abandonned. I hope this gives you the context.

Here is what is see in the `stats/` directory and my recommendation. @saguilarDevel @adjih @MarinoMtz I'm looking for your input on whether you agree with this approach:
- `log.txt` and `log2.txt` are artefact which should never have been committed. 
- `tao_calculator.py` contains only a single `get_toa()` function. I suggest to create a module called `OCUtils.py` and put that function there (as well as other generic utility functions we encounter later on)
- similarly `cdf_calc.py` contains only `cdf_cal()`, I recommend to add it to `OpenSchcUtils.py`
- `statsct.py` is a proper module, and defines teh `Statsct` class. I recommend to move it to `src/` and rename the file and class to `OCStats`. What does "ct" stand for in  `statsct.py`

**Comment by adjih** [comment]
This looks good to me. 

Minor comments about the naming, basing on previous renaming in hackathons and a few prior conventions:
- possibly we could keep lower case for module names in `src/` (as per PEP 8)
- previous renaming in `src/` was using prefix `gen_` for generic, `net_` for network, `net_sim` for simulator, `compr_` for de/compression, `frag_` from de/fragmentation code, etc.

But I am fine with what is proposed, the key point is to remove `stats/` directory

**Comment by twatteyne** [comment]
@adjih thanks for the context about naming! I propose to put tha in the `README.md`, I created #123.

Why not having all the simulator related things start with `sim_`?!?

**Comment by twatteyne** [comment]
I know what to do here, I'm taking ownership over this issue and will open a PR shortly.
